kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/internal/KafkaProxyFrontendHandler.java

```diff
@@ -311,29 +478,79 @@ public void initiateConnect(HostPort remote, List<FilterAndInvoker> filters) {
             pipeline.addFirst("networkLogger", new LoggingHandler("io.kroxylicious.proxy.internal.UpstreamNetworkLogger"));
         }
 
-        virtualCluster.getUpstreamSslContext().ifPresent(c -> pipeline.addFirst("ssl", c.newHandler(outboundChannel.alloc(), remote.host(), remote.port())));
-
-        connectFuture.addListener(future -> {
+        virtualCluster.getUpstreamSslContext().ifPresent(c -> {
+            final SslHandler handler = c.newHandler(outboundChannel.alloc(), remote.host(), remote.port());
+            pipeline.addFirst("ssl", handler);
+            sslHandshakeFuture.set(handler.handshakeFuture());
+            registerExceptionResponse(SSLHandshakeException.class, throwable -> {
```

Comment:
Are we really only interested in the SSLHandshakeException?  Should we handle a "Connection refused" in the same way?

I wonder if the logic should be:

```
if we have accumulated frames in bufferedMsgs
   buildErrorResponseFrame based on first frame
end if
 log exception
 close connection
```

Reply:
> I think there are a whole lot of exceptions we would want to do something like for. In this PR I was just wanting to fix the SSL error and bootstrap the approach.

Ok.  My gut feel is that we just don't need to the exception check, but you are much closer to the problem than me!

> I didn't want to log "anticipatable" exceptions due to concerns about log spam. But I do also want to add metrics around uncaught exceptions and probably failed connections.

I think we will have to log the upstream connection error.  If a broker cert expires, we need something for admins to find.    They won't thank us if all they have is a BrokerNotAvailableException/NetworkException on the application side and nothing proxy side to reveal the actual issue.

Reply:
We should raise an issue to add `upstreamRefusesConnection` IT test.

Reply:
done https://github.com/kroxylicious/kroxylicious/issues/1493